internal/datanode/flow_graph_manager.go

```diff
@@ -0,0 +1,127 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datanode
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+
+	"go.uber.org/zap"
+)
+
+type flowgraphManager struct {
+	flowgraphs sync.Map // vChannelName -> dataSyncService
+}
+
+func newFlowgraphManager() *flowgraphManager {
+	return &flowgraphManager{}
+}
+
+func (fm *flowgraphManager) AddAndStart(dn *DataNode, vchan *datapb.VchannelInfo) error {
+	log.Debug("Received Vchannel Info",
+		zap.String("vChannelName", vchan.GetChannelName()),
+		zap.Int("Unflushed Segment Number", len(vchan.GetUnflushedSegments())),
+		zap.Int("Flushed Segment Number", len(vchan.GetFlushedSegments())),
+	)
+
+	if _, ok := fm.flowgraphs.Load(vchan.GetChannelName()); ok {
+		log.Warn("try to add an existed DataSyncService", zap.String("vChannelName", vchan.GetChannelName()))
+		return nil
+	}
+
+	replica, err := newReplica(dn.ctx, dn.rootCoord, vchan.GetCollectionID())
+	if err != nil {
+		log.Warn("newReplica failed", zap.String("vChannelName", vchan.GetChannelName()), zap.Error(err))
+		return err
+	}
+
+	var alloc allocatorInterface = newAllocator(dn.rootCoord)
+
+	dataSyncService, err := newDataSyncService(dn.ctx, make(chan flushMsg, 100), replica, alloc, dn.msFactory, vchan, dn.clearSignal, dn.dataCoord, dn.segmentCache, dn.blobKv, dn.compactionExecutor)
+	if err != nil {
+		log.Warn("newDataSyncService fail", zap.String("vChannelName", vchan.GetChannelName()), zap.Error(err))
+		return err
+	}
+	log.Info("successfully created dataSyncService", zap.String("vChannelName", vchan.GetChannelName()))
```

Comment:
If the `dataSyncService` is not proper noun, we can just print `data sync service` or ` service to sync data` to make it clear.